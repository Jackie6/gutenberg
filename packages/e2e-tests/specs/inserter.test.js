/**
 * External dependencies
 */
import { sortBy, uniq } from 'lodash';

/**
 * WordPress dependencies
 */
import {
	createNewPost,
	insertBlock,
	openGlobalInserter,
} from '@wordpress/e2e-test-utils';

describe( 'Inserter', () => {
	const openAllInserterCategories = async () => {
		const notOpenCategoryPanels = await page.$$(
			'.editor-inserter__results .components-panel__body:not(.is-opened)'
		);
		for ( const categoryPanel of notOpenCategoryPanels ) {
			await categoryPanel.click();
		}
	};

	const getAllInserterItemTitles = async () => {
		const inserterItemTitles = await page.evaluate( () => {
			return Array.from(
				document.querySelectorAll(
					'.editor-inserter__results .editor-block-types-list__item-title'
				)
			).map(
				( inserterItem ) => {
					return inserterItem.innerText;
				}
			);
		} );
		return sortBy( uniq( inserterItemTitles ) );
	};

	beforeEach( async () => {
		await createNewPost();
	} );

	it( 'all blocks should appear', async () => {
		await page.click( '.editor-default-block-appender' );
		await openGlobalInserter();
		await openAllInserterCategories();
		expect( await getAllInserterItemTitles() ).toMatchSnapshot();
	} );

	it( 'media & text block should restrict allowed blocks', async () => {
		await insertBlock( 'Media & Text' );
		await page.click( '.wp-block-media-text .editor-rich-text' );
		await openGlobalInserter();
		await openAllInserterCategories();
		expect( await getAllInserterItemTitles() ).toMatchSnapshot();
	} );

	it( 'no blocks should be allowed inside columns', async () => {
		await insertBlock( 'Columns' );
		await page.click( '[aria-label="Block Navigation"]' );
		const columnBlockMenuItem = ( await page.$x( '//button[contains(concat(" ", @class, " "), " editor-block-navigation__item-button ")][text()="Column"]' ) )[ 0 ];
		await columnBlockMenuItem.click();
		await openGlobalInserter();
		await openAllInserterCategories();
		expect( await getAllInserterItemTitles() ).toHaveLength( 0 );
	} );
} );
