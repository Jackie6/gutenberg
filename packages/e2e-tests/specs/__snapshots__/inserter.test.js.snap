// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inserter all blocks should appear 1`] = `
Array [
  "Amazon Kindle",
  "Animoto",
  "Archives",
  "Audio",
  "Button",
  "Calendar",
  "Categories",
  "Classic",
  "Cloudup",
  "Code",
  "CollegeHumor",
  "Columns",
  "Cover",
  "Crowdsignal",
  "Custom HTML",
  "Dailymotion",
  "Embed",
  "Facebook",
  "File",
  "Flickr",
  "Gallery",
  "Heading",
  "Hulu",
  "Image",
  "Imgur",
  "Inline Image",
  "Instagram",
  "Issuu",
  "Kickstarter",
  "Latest Comments",
  "Latest Posts",
  "List",
  "Media & Text",
  "Meetup.com",
  "Mixcloud",
  "More",
  "Page Break",
  "Paragraph",
  "Preformatted",
  "Pullquote",
  "Quote",
  "RSS",
  "Reddit",
  "ReverbNation",
  "Screencast",
  "Scribd",
  "Search",
  "Separator",
  "Shortcode",
  "Slideshare",
  "SmugMug",
  "SoundCloud",
  "Spacer",
  "Speaker Deck",
  "Spotify",
  "TED",
  "Table",
  "Tag Cloud",
  "Tumblr",
  "Twitter",
  "Verse",
  "Video",
  "VideoPress",
  "Vimeo",
  "WordPress",
  "WordPress.tv",
  "YouTube",
]
`;

exports[`Inserter media & text block should restrict allowed blocks 1`] = `
Array [
  "Button",
  "Heading",
  "Inline Image",
  "List",
  "Paragraph",
]
`;
